### PR TITLE
Express pattern guards using Jane_syntax

### PR DESCRIPTION
**Pull request by rajgodse**
Changing `Parsetree` requires painful updates to ppx's, so this PR rewrites the pattern guard feature using `Jane_syntax`.

In doing so, it creates a new syntactic category: `Jane_syntax.Case`.

The attribute marking the extension and the location of a `Parsetree.case` which represents a `Jane_syntax.Case.t` are stored in its field `pc_rhs: Parsetree.expression`. This rhs should **never** be interpreted as an expression by itself, which is why everywhere in the code that accesses the rhs of a `Parsetree.case` should first apply `Jane_syntax.Case.of_ast` (as documented in `parsetree.mli`). Ensuring that this invariant is introduced everywhere is one of the main difficulties of this PR.

However, the field `pc_rhs` is rarely accessed: ripgrepping yields the following files:
```
printer/printast_with_mappings.ml
ocaml/typing/typecore.ml
ocaml/tools/ocamlprof.ml
ocaml/typing/untypeast.ml
ocaml/parsing/parsetree.mli
ocaml/parsing/depend.ml
ocaml/parsing/jane_syntax.ml
ocaml/parsing/ast_iterator.ml
ocaml/parsing/jane_syntax_parsing.ml
ocaml/parsing/ast_helper.ml
ocaml/parsing/printast.ml
ocaml/parsing/pprintast.ml
ocaml/parsing/ast_mapper.ml
```

Of these, only `printast.ml` doesn't apply `of_ast` for cases, but this is fine as it doesn't apply it for any other syntactic category (it just prints the raw Parsetree instead).

Currently, there is only one constructor for the variant `Jane_syntax.Case.t`, namely `Jcase_pattern_guarded of Jane_syntax.Pattern_guarded.case`. The payload of this constructor stores the same data as `Ppattern_guarded_rhs` did previously, together with the case's lhs.

It is also worth noting the exact ast representation of `Jane_syntax.Pattern_guarded.Pg_case { scrutinee; cases }` is `case : Parsetree.case` with `case.pc_rhs.pexp_desc = Pexp_match (scrutinee, cases)`, with extension attribute and location stored as described above.

### CODE REVIEW

File: ocaml/testsuite/tests/pattern-guards/test.ml
Comment (ncik-roberts) [inline ocaml/testsuite/tests/pattern-guards/test.ml:739]: I think we can get rid of this principality warning if we move the match on `y` to be inside the case with the `when` guard, as opposed to an entirely unrelated case.

File: ocaml/parsing/jane_syntax.mli
Comment (ncik-roberts) [inline ocaml/parsing/jane_syntax.mli:260]: This attribute list is always empty, no? I think we should probably just return `t`, and not `t` together with attributes.

File: ocaml/parsing/jane_syntax_parsing.ml
Comment (ncik-roberts) [inline ocaml/parsing/jane_syntax_parsing.ml:663]: This language extension is not "erasable" &mdash; we're not easily going to be able to rewrite programs that use this feature into ones that don't simply by dropping attributes. For that reason, we have more flexibility with our choice of embedding. In particular, we can use extensions instead of attributes as our means of embedding.

If we do this, an uninterpreted extension node will be wrapped around the expression (instead of just adding attributes) so, if the extension node somehow makes it through compilation, the middle end will raise an exception instead of silently miscompiling.

**Review by ncik-roberts — COMMENTED** [review]
I haven't finished review, but I've marked the files I have finished reviewing as "Viewed" and will continue from there the next time I take a look.

File: ocaml/parsing/parsetree.mli
Comment (ncik-roberts) [inline ocaml/parsing/parsetree.mli:440]: I don't like this advice &mdash; let's try to keep details of the embedding outside of this file. More generally, I'd prefer to just call `of_ast` everywhere because it's pretty simple and correct.

  1. Can we delete this part of the comment?
  2. Can we change any place that currently takes this advice to instead call `of_ast`?

File: ocaml/parsing/parsetree.mli
Comment (ncik-roberts) [inline ocaml/parsing/parsetree.mli:436]:
```suggestion
    Use [Jane_syntax.Case.of_ast] before using these fields
    directly, as the former will detect language extensions correctly.
```

File: ocaml/testsuite/tests/parsetree/source.ml
Comment (ncik-roberts) [inline ocaml/testsuite/tests/parsetree/source.ml:7481]: Are these tests interesting in any way? Should we try to preserve them in the intermediate world where this feature is not yet sent upstream?

File: ocaml/tools/ocamlprof.ml
Comment (ncik-roberts) [inline ocaml/tools/ocamlprof.ml:378]: One goal of this PR is to reduce the diff with upstream &mdash; the total number of lines changed by this branch should go down. Can you spot-check the overall diff between the base of the `pattern-guards` branch and this commit to see if that in fact pans out?

File: ocaml/typing/typecore.ml
Comment (ncik-roberts) [inline ocaml/typing/typecore.ml:3523]: I don't agree with the rename from `case_rhs` to `case` here &mdash; there is still a well-defined notion of "case rhs" on the AST, it's just that the embedding is a bit more complicated.

File: ocaml/typing/typecore.ml
Comment (ncik-roberts) [inline ocaml/typing/typecore.ml:3789]: As discussed, let's introduce the `Psimple_rhs ... | ...` type into this file so we can minimize the diff.

File: ocaml/typing/typecore.ml
Comment (ncik-roberts) [inline ocaml/typing/typecore.ml:7045]: I was somewhat surprised at how large this diff is. (I mean, I think I see that the diff is because of reordering/indentation rather than bona fide code changes.) It would be nice to pare it down a bit more, if that's possible. (It might not be.)

**Review by ncik-roberts — COMMENTED** [review]
Done revieweing for now. I made a few minor comments.

File: ocaml/testsuite/tests/pattern-guards/test.ml
Comment (rajgodse) [reply ocaml/testsuite/tests/pattern-guards/test.ml]: I think moving it inside the associated case actually tests the thing we want, but even prior to this change, it would issue a principality warning.

File: ocaml/parsing/jane_syntax_parsing.ml
Comment (rajgodse) [reply ocaml/parsing/jane_syntax_parsing.ml]: Makes sense! The new encoding applies the extension node to the case rhs, let me know if this doesn't make sense.

File: ocaml/parsing/jane_syntax_parsing.mli
Comment (rajgodse) [reply ocaml/parsing/jane_syntax_parsing.mli:190]: The reason is that when we call `Case.of_ast` on a normal (say, unguarded) case whose rhs happens to use a Jane_syntax extension, we don't want `Case.of_ast` to fail (which it would absent this flag being set to false).

Let me know if this is confusing or if you have a better way to design this.

File: ocaml/parsing/jane_syntax_parsing.ml
Comment (rajgodse) [reply ocaml/parsing/jane_syntax_parsing.ml]: Good catch, done!

File: ocaml/testsuite/tests/parsetree/source.ml
Comment (rajgodse) [reply ocaml/testsuite/tests/parsetree/source.ml:7481]: As I understand, the purpose of this file is to test the parser and source printer. I'm not sure if these tests are particularly interesting, but we had a reason for adding them in #1588: do you remember what that was?

File: ocaml/parsing/parser.mly
Comment (rajgodse) [reply ocaml/parsing/parser.mly]: works for me

File: ocaml/parsing/parsetree.mli
Comment (rajgodse) [reply ocaml/parsing/parsetree.mli]: I don't believe there is any place in the code that takes this advice, but it is the advice given when mentioning the other Jane_syntax syntactic categories. I agree that it is better to always call `of_ast`, so I deleted this part of the comment.

File: ocaml/typing/typecore.ml
Comment (rajgodse) [reply ocaml/typing/typecore.ml]: Resolved by the below

File: ocaml/typing/typecore.ml
Comment (rajgodse) [reply ocaml/typing/typecore.ml]: Done, although we can further reduce the diff by using the original field names. I opted to omit the prefixes (like `pbg_`) for now. What do you think?

File: ocaml/typing/typecore.ml
Comment (rajgodse) [reply ocaml/typing/typecore.ml]: Resolved by the above

File: ocaml/tools/ocamlprof.ml
Comment (rajgodse) [reply ocaml/tools/ocamlprof.ml:378]: The number of changes actually looks pretty similar... do you think it's worth going through the diff with the branch base and manually reducing the number of changes when possible?

File: ocaml/parsing/jane_syntax_parsing.mli
Comment (ncik-roberts) [reply ocaml/parsing/jane_syntax_parsing.mli:190]: As discussed, let's move this to the `AST_internal` module instead so we can co-locate this boolean flag with the reason why we need it.

File: ocaml/parsing/ast_helper.mli
Comment (ncik-roberts) [inline ocaml/parsing/ast_helper.mli:193]: The original `ast_helper.mli` had:

```ocaml
val case : pattern -> ?guard:expression -> expression -> case
```

(See https://github.com/ocaml-flambda/flambda-backend/blame/1b6682660307f054113331959b7a064246a7a60b/ocaml/parsing/ast_helper.mli#L193.)

I'd prefer if this PR more consciously reverted the API to its state prior to your branch. (E.g. compiler libraries are included as part of the distribution, so user code can call functions like `Ast_helper.case`.)

File: ocaml/parsing/jane_syntax.ml
Comment (ncik-roberts) [inline ocaml/parsing/jane_syntax.ml:358]:
```suggestion
        Pg_case { lhs = pc_lhs; scrutinee; cases }
```

File: ocaml/typing/typecore.ml
Comment (ncik-roberts) [inline ocaml/typing/typecore.ml:824]: How about renaming to `parsed_case_rhs_of_case`? I don't think we should abbreviate unless the compactness is for some reason clarifying.

File: ocaml/typing/typecore.ml
Comment (ncik-roberts) [inline ocaml/typing/typecore.ml:3764]: I think this could be a bug: we're extracting out `pc_rhs` without checking whether it's Jane Syntax. `type_approx_*` is used for "approximately" type-checking all the bindings in a recursive knot before type-checking them for real &mdash; it's probably worth adding a test that involves `let rec f = function ...` where the function has a when-guard to see if this is currently a bug.

**Review by ncik-roberts — COMMENTED** [review]
Thanks, this is easier to review now.

File: ocaml/parsing/ast_helper.mli
Comment (rajgodse) [reply ocaml/parsing/ast_helper.mli]: Done for this specific case. I will take a look at the overall diff to try to revert any APIs that don't need to be modified (and generally minimize the size of the diff).

File: ocaml/parsing/jane_syntax_parsing.ml
Comment (rajgodse) [reply ocaml/parsing/jane_syntax_parsing.ml]: works for me!

File: ocaml/typing/typecore.ml
Comment (rajgodse) [reply ocaml/typing/typecore.ml:3764]: Added such a test, which passes on this version of the code (although because I don't know what the result of `type_approx_*` is used for, it could be silently producing the wrong result).

I still suspect that the code is wrong though, so I fixed the bug (I believe) by instead getting out a parsed rhs and passing it around.

File: ocaml/testsuite/tests/parsetree/source.ml
Comment (rajgodse) [reply ocaml/testsuite/tests/parsetree/source.ml:7481]: I now remember: these tests check that parsing and source dumping roundtrips. This remains important, right?

File: ocaml/parsing/ast_helper.mli
Comment (rajgodse) [reply ocaml/parsing/ast_helper.mli]: Did this!

File: ocaml/parsing/jane_syntax_parsing.ml
Comment (ncik-roberts) [inline ocaml/parsing/jane_syntax_parsing.ml:679]: This wants a comment explaining why it's set to `false` where every other one is set to `true`.

File: ocaml/parsing/jane_syntax_parsing.ml
Comment (ncik-roberts) [reply ocaml/parsing/jane_syntax_parsing.ml:475]: You're welcome to edit more. I found the old wording confusing. (I recognize that bits of the comment were transplanted from elsewhere.)

**Review by ncik-roberts — COMMENTED** [review]
I have a few small comments, but nothing major. This is coming along well &mdash; we should be able to merge soon.

File: ocaml/parsing/jane_syntax_parsing.ml
Comment (ncik-roberts) [reply ocaml/parsing/jane_syntax_parsing.ml:679]: It's possible that this should just reference the mli, if you end up including the commented I suggested for that.

File: ocaml/testsuite/tests/parsetree/source.ml
Comment (rajgodse) [reply ocaml/testsuite/tests/parsetree/source.ml:7481]: Good call, done!

(Upon initially reading this comment, I didn't realize that the PR deletes these tests)

File: ocaml/parsing/jane_syntax.ml
Comment (rajgodse) [reply ocaml/parsing/jane_syntax.ml:354]: Done!

File: ocaml/parsing/jane_syntax_parsing.ml
Comment (rajgodse) [reply ocaml/parsing/jane_syntax_parsing.ml:690]: Added a comment referencing the previous one.

**Comment by rajgodse** [comment]
Builds will fail until #1747 is merged.

**Review by ncik-roberts — APPROVED** [review]
This will be good to merge once it turns green.

**Pull request merged**
